### PR TITLE
[RFC] core: make pager aliased paged not always writable

### PR DESCRIPTION
**Pull request by etienne-lms**
This change lowers the attack surface of executable memory in the pager by allowing write access to aliased virtual pages related to read-only content (including executable content) only when pager needs to update page content.

### CODE REVIEW

**Comment by jforissier** [comment]
Won't this have a big performance impact? (due to the TLB flush)

File: core/arch/arm/mm/tee_pager.c [hunk 1]
```diff
@@ -260,13 +260,18 @@ static void *pager_add_alias_page(paddr_t pa)
...
 			(TEE_MATTR_CACHE_CACHED << TEE_MATTR_CACHE_SHIFT) |
-			TEE_MATTR_SECURE | TEE_MATTR_PRW;
+			TEE_MATTR_SECURE | TEE_MATTR_PR;
 
 	DMSG("0x%" PRIxPA, pa);
 
 	if (!pager_alias_next_free || !ti->num_entries)
 		panic("invalid alias entry");
 
+#ifndef CFG_CORE_RWDATA_NOEXEC
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/tee_pager.c]: I'd rather try to protect as much as possible in the pager. So if no one objects please remove all these ifdefs in this patch.

**Comment by jenswi-linaro** [comment]
I don't think you'll notice it. But I think we should to something about all those TODOs, that could actually improve the performance a bit.

**Comment by etienne-lms** [comment]
> Won't this have a big performance impact? (due to the TLB flush)

I have tested on my HW and impact was quite slow, but I did not test with a stressed setup on a real usecase... That said, pager already flushes TLB when updating the mmu. Maybe this would deserve some benchmark tests :)

**Comment by jforissier** [comment]
@jenswi-linaro @etienne-lms OK, I suppose we can deal with that in a separate patch.

File: core/arch/arm/mm/tee_pager.c (hunk 1 above)
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c]: I do not object. But your suggestion tends me to believe even `CFG_CORE_RWDATA_NOEXEC` configuration switch should be discarded and related code always embedded.

Comment (jenswi-linaro) [inline core/arch/arm/mm/tee_pager.c]: Fine with me to remove `CFG_CORE_RWDATA_NOEXEC`.

Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c]: Ok. I propose that (enforce seperation of rx and rw memory) in a specific commit, hence a specifc PR. Meanwhile, i'll update this PR to remove this condition in the pager only.

File: core/arch/arm/mm/tee_pager.c [hunk 2]
```diff
@@ -484,6 +485,21 @@ static void tee_pager_load_page(struct tee_pager_area *area, vaddr_t page_va,
...
+
+	/* Insure we are allowed to write to aliased virtual page */
+	ti = &pager_alias_tbl_info;
+	idx_alias = core_mmu_va2idx(ti, (vaddr_t)va_alias);
+	core_mmu_get_entry(ti, idx_alias, &pa_alias, &attr_alias);
+	if (!(attr_alias & TEE_MATTR_PW)) {
+		attr_alias |= TEE_MATTR_PW;
+		core_mmu_set_entry(ti, idx_alias, pa_alias, attr_alias);
+		/* TODO: flush TLB for target page only */
+		core_tlb_maintenance(TLBINV_UNIFIEDTLB, 0);
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/tee_pager.c]: If we always made sure that aliased pages aren't accessible unless in this function there wouldn't be any need for this flush. But that's something for another PR.

File: core/arch/arm/mm/tee_pager.c [hunk 3]
```diff
@@ -484,6 +485,21 @@ static void tee_pager_load_page(struct tee_pager_area *area, vaddr_t page_va,
 {
 	size_t idx = (page_va - area->base) >> SMALL_PAGE_SHIFT;
 	const void *stored_page = area->store + idx * SMALL_PAGE_SIZE;
+	struct core_mmu_table_info *ti __maybe_unused;
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/tee_pager.c]: The `__maybe_unused` attributes can be removed now.

File: core/arch/arm/mm/tee_pager.c (hunk 2 above)
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c]: > made sure that aliased pages aren't accessible unless in this function

Any idea ? Seems hard to implement. Through kasan?

**Comment by etienne-lms** [comment]
Comments addressed.

This change protects alias pages related the read-only (potentially read-exec) memory. Should it be extended to read/write memory ? Such should also not be accessed from aliased vmem.

Travis failure due to issue in optee_client, fixed through https://github.com/OP-TEE/optee_client/pull/91.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
Tags applied.
Travis should make me happy now.

**Comment by etienne-lms** [comment]
Travis complains on trace log size. Has something changed about it ? Or is my PR buggy in some way ?

**Comment by jforissier** [comment]
> Travis complains on trace log size. Has something changed about it ?

No, it's just that `make check` (QEMU xtest) failed, and in this case we dump the debug traces which is a lot of text.

> Or is my PR buggy in some way ?

I don't know, in the raw log we can see that the failing test case is:
```
o regression_5006.16 Allocate_sharedMemory_MaxSize_Above_and_Below, allocate max size
/home/travis/optee_repo/optee_test/host/xtest/regression_5000.c:443: AllocateSharedMemory(&cs->context, &shm, size_max, 0x00000001) has an unexpected value: 0xffff000c = TEEC_ERROR_OUT_OF_MEMORY, expected 0x0 = TEEC_SUCCESS
  regression_5006.16 FAILED
```

**Comment by etienne-lms** [comment]
ok, something about the last changes merged in optee_client.

**Comment by jforissier** [comment]
@etienne-lms yeah, the solution is probably to remove the tests that are trying to allocate TEEC_CONFIG_SHAREDMEM_MAX_SIZE more or less.

**Comment by etienne-lms** [comment]
rebased. travis at work.

**Comment by etienne-lms** [comment]
rebased.

**Pull request merged**